File: worker/app/blueprint_contracts/trade_catalog.py
"""Trade catalog contract (Python twin). Side-effect free.

Source of truth: docs/blueprint-trade-catalog.md
"""
from __future__ import annotations

from typing import Iterable, Literal, Optional, TypedDict

TradeSupportStatus = Literal[
    "mvp_supported", "measurement_object_only", "future_supported", "unsupported"
]

MVP_SUPPORTED_TRADES: tuple[str, ...] = (
    "roofing",
    "exterior_walls_siding",
    "paint_coatings",
    "gutters_fascia_trim",
)
MEASUREMENT_OBJECT_ONLY_TRADES: tuple[str, ...] = ("windows_doors",)
FUTURE_SUPPORTED_TRADES: tuple[str, ...] = (
    "drywall",
    "framing",
    "insulation",
    "flooring",
    "concrete",
    "electrical",
    "plumbing",
    "hvac",
)
UNSUPPORTED_TRADES: tuple[str, ...] = ()

TRADE_SUPPORT_MAP: dict[str, TradeSupportStatus] = {
    "roofing": "mvp_supported",
    "exterior_walls_siding": "mvp_supported",
    "paint_coatings": "mvp_supported",
    "gutters_fascia_trim": "mvp_supported",
    "windows_doors": "measurement_object_only",
    "drywall": "future_supported",
    "framing": "future_supported",
    "insulation": "future_supported",
    "flooring": "future_supported",
    "concrete": "future_supported",
    "electrical": "future_supported",
    "plumbing": "future_supported",
    "hvac": "future_supported",
}


class AcceptTradeResult(TypedDict, total=False):
    ok: bool
    reason: str
    flag_code: str


def validate_trade_support_status(value: str) -> bool:
    return value in {
        "mvp_supported",
        "measurement_object_only",
        "future_supported",
        "unsupported",
    }


def _status(trade_id: str) -> Optional[TradeSupportStatus]:
    return TRADE_SUPPORT_MAP.get(trade_id)


def is_mvp_supported_trade(trade_id: str) -> bool:
    return _status(trade_id) == "mvp_supported"


def is_measurement_object_only_trade(trade_id: str) -> bool:
    return _status(trade_id) == "measurement_object_only"


def is_future_supported_trade(trade_id: str) -> bool:
    return _status(trade_id) == "future_supported"


def assert_can_accept_trade_for_mvp(
    trade_id: str,
    accepted_trade_ids_in_session: Iterable[str],
    review_state: Optional[str] = None,
) -> AcceptTradeResult:
    status = _status(trade_id)
    accepted = set(accepted_trade_ids_in_session)

    if status == "measurement_object_only":
        return {
            "ok": False,
            "reason": "windows_doors is measurement-object-only and cannot be a top-level accepted trade",
            "flag_code": "windows_doors_selected_as_trade",
        }
    if status == "unsupported" or status is None:
        return {
            "ok": False,
            "reason": "trade is unsupported",
            "flag_code": "unsupported_trade_for_mvp",
        }
    if status == "future_supported":
        if review_state != "manual_only":
            return {
                "ok": False,
                "reason": "future_supported trades require manual_only review_state during MVP",
                "flag_code": "future_trade_requires_sheet_intelligence",
            }
        return {"ok": True}
    if trade_id == "paint_coatings" and "exterior_walls_siding" not in accepted:
        return {
            "ok": False,
            "reason": "paint_coatings is derived from exterior_walls_siding and cannot stand alone",
            "flag_code": "paint_without_wall_source",
        }
    return {"ok": True}

File: supabase/functions/_shared/blueprint-importer/trade-catalog.ts
// Blueprint Importer v2 — Trade catalog contract (Phase 1, side-effect-free).
// Source of truth: docs/blueprint-trade-catalog.md
// DO NOT call DB / APIs / extraction / estimate generation from this file.

export type TradeSupportStatus =
  | "mvp_supported"
  | "measurement_object_only"
  | "future_supported"
  | "unsupported";

export type TradeId =
  | "roofing"
  | "exterior_walls_siding"
  | "paint_coatings"
  | "gutters_fascia_trim"
  | "windows_doors"
  | "drywall"
  | "framing"
  | "insulation"
  | "flooring"
  | "concrete"
  | "electrical"
  | "plumbing"
  | "hvac";

export const MVP_SUPPORTED_TRADES: readonly TradeId[] = [
  "roofing",
  "exterior_walls_siding",
  "paint_coatings",
  "gutters_fascia_trim",
] as const;

export const MEASUREMENT_OBJECT_ONLY_TRADES: readonly TradeId[] = [
  "windows_doors",
] as const;

export const FUTURE_SUPPORTED_TRADES: readonly TradeId[] = [
  "drywall",
  "framing",
  "insulation",
  "flooring",
  "concrete",
  "electrical",
  "plumbing",
  "hvac",
] as const;

export const UNSUPPORTED_TRADES: readonly TradeId[] = [] as const;

export const TRADE_SUPPORT_MAP: Readonly<Record<TradeId, TradeSupportStatus>> = {
  roofing: "mvp_supported",
  exterior_walls_siding: "mvp_supported",
  paint_coatings: "mvp_supported",
  gutters_fascia_trim: "mvp_supported",
  windows_doors: "measurement_object_only",
  drywall: "future_supported",
  framing: "future_supported",
  insulation: "future_supported",
  flooring: "future_supported",
  concrete: "future_supported",
  electrical: "future_supported",
  plumbing: "future_supported",
  hvac: "future_supported",
};

export function isMvpSupportedTrade(trade: TradeId): boolean {
  return TRADE_SUPPORT_MAP[trade] === "mvp_supported";
}
export function isMeasurementObjectOnlyTrade(trade: TradeId): boolean {
  return TRADE_SUPPORT_MAP[trade] === "measurement_object_only";
}
export function isFutureSupportedTrade(trade: TradeId): boolean {
  return TRADE_SUPPORT_MAP[trade] === "future_supported";
}

/**
 * Phase 0 rule enforcement: returns null if the trade is acceptable in the MVP
 * acceptance flow, or a string reason if it must be blocked. Pure / deterministic.
 *
 * - windows_doors: measurement-object-only — never a top-level accepted MVP trade.
 * - paint_coatings: derived — requires exterior_walls_siding accepted in same session.
 * - future_supported: allowed only with manual_only review state.
 * - unsupported: always blocked.
 */
export function assertCanAcceptTradeForMvp(input: {
  trade_id: TradeId;
  accepted_trade_ids_in_session: TradeId[];
  review_state?: "pending_review" | "blocked" | "cleared" | "manual_only";
}): { ok: true } | { ok: false; reason: string; flag_code: string } {
  const status = TRADE_SUPPORT_MAP[input.trade_id];

  if (status === "measurement_object_only") {
    return {
      ok: false,
      reason: "windows_doors is measurement-object-only and cannot be a top-level accepted trade",
      flag_code: "windows_doors_selected_as_trade",
    };
  }
  if (status === "unsupported") {
    return { ok: false, reason: "trade is unsupported", flag_code: "unsupported_trade_for_mvp" };
  }
  if (status === "future_supported") {
    if (input.review_state !== "manual_only") {
      return {
        ok: false,
        reason: "future_supported trades require manual_only review_state during MVP",
        flag_code: "future_trade_requires_sheet_intelligence",
      };
    }
    return { ok: true };
  }
  if (input.trade_id === "paint_coatings") {
    if (!input.accepted_trade_ids_in_session.includes("exterior_walls_siding")) {
      return {
        ok: false,
        reason: "paint_coatings is derived from exterior_walls_siding and cannot stand alone",
        flag_code: "paint_without_wall_source",
      };
    }
  }
  return { ok: true };
}
